From 815160bcbea62e1b2c1a04e4acb1e55474d6e90f Mon Sep 17 00:00:00 2001
From: Sidhanta Sahu <quic_sidhanta@quicinc.com>
Date: Thu, 6 Mar 2025 00:27:29 -0800
Subject: [PATCH 1/3] UPSTREAM:CP : wifi: ath12k: Add support to create primary
 peers list

Add support to retrieve primary link peers from a partner peer.
Currently, all peer objects per SoC are maintained in ath12k_base's
peer-related list. Identifying the primary peer list requires traversing
the ath12k_hw_group object, accessing its ath12k_base objects, and
looping through all ath12k_peers. This process necessitates a mutex
lock, which cannot be acquired in certain contexts like telemetry
workqueue fetching stats while holding a spin_lock.

Introduce a mechanism to maintain a list of primary link peers to
address this limitation.

Patch-dependency: None
Patch-work: None

Signed-off-by: Sidhanta Sahu <quic_sidhanta@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/peer.c | 145 ++++++++++++++++++++++++-
 drivers/net/wireless/ath/ath12k/peer.h |   8 ++
 2 files changed, 151 insertions(+), 2 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/peer.c
+++ b/drivers/net/wireless/ath/ath12k/peer.c
@@ -12,6 +12,139 @@
 #endif
 #include "test_mlo_mesh.h"
 
+static LIST_HEAD(ath12k_primary_peers);
+static DEFINE_SPINLOCK(ath12k_primary_peers_lock);
+
+void ath12k_peer_add_primary(struct ath12k_peer *peer)
+{
+	/* Must call when a peer is identified as primary */
+	if (!peer || !peer->primary_link)
+		return;
+
+	INIT_LIST_HEAD(&peer->primary_list);
+
+	spin_lock(&ath12k_primary_peers_lock);
+	ath12k_dbg(NULL, ATH12K_DBG_PEER,
+		   "peer: %pM added to primary list\n", peer->addr);
+	list_add_rcu(&peer->primary_list, &ath12k_primary_peers);
+	spin_unlock(&ath12k_primary_peers_lock);
+}
+
+void ath12k_peer_remove_primary(struct ath12k_peer *peer)
+{
+	if (!peer || !peer->primary_link)
+		return;
+
+	spin_lock(&ath12k_primary_peers_lock);
+	ath12k_dbg(NULL, ATH12K_DBG_PEER,
+		   "peer: %pM removed from primary list\n", peer->addr);
+	if (!list_empty(&peer->primary_list))
+		list_del_rcu(&peer->primary_list);
+
+	INIT_LIST_HEAD(&peer->primary_list);
+	spin_unlock(&ath12k_primary_peers_lock);
+}
+
+void ath12k_peer_purge_primary_list(void)
+{
+	spin_lock(&ath12k_primary_peers_lock);
+	INIT_LIST_HEAD(&ath12k_primary_peers);
+	spin_unlock(&ath12k_primary_peers_lock);
+
+	synchronize_rcu();
+}
+
+static void ath12k_peer_rcu_callback(struct rcu_head *rcu)
+{
+	struct ath12k_peer *peer = container_of(rcu, struct ath12k_peer, rcu);
+
+	kfree(peer);
+	ath12k_dbg(NULL, ATH12K_DBG_PEER, "peer freed\n");
+}
+
+void ath12k_peer_purge_primary_list_per_ab(struct ath12k_base *ab)
+{
+	struct ath12k_peer *peer;
+
+	if (!ab)
+		return;
+
+	lockdep_assert_held(&ab->base_lock);
+	list_for_each_entry(peer, &ab->peers, list) {
+		if (!peer->primary_link)
+			continue;
+
+		spin_lock(&ath12k_primary_peers_lock);
+		if (!list_empty(&peer->primary_list)) {
+			list_del_rcu(&peer->primary_list);
+			ath12k_dbg(NULL, ATH12K_DBG_PEER,
+				   "peer: %pM removed from primary list\n",
+				   peer->addr);
+		}
+		spin_unlock(&ath12k_primary_peers_lock);
+		INIT_LIST_HEAD(&peer->primary_list);
+	}
+}
+
+void ath12k_peer_dump_primary_peers(void)
+{
+	struct ath12k_peer *peer;
+
+	rcu_read_lock();
+
+	list_for_each_entry_rcu(peer, &ath12k_primary_peers, primary_list) {
+		if (peer && peer->sta) {
+			ath12k_dbg(NULL, ATH12K_DBG_PEER,
+				   "dump peer: %pM sta: %pM \n",
+				   peer->addr, peer->sta->addr);
+		}
+	}
+
+	rcu_read_unlock();
+}
+
+struct ath12k_peer* ath12k_peer_find_primary_by_sta_addr(const u8 *sta_addr)
+{
+	struct ath12k_peer *found_peer = NULL;
+	struct ath12k_peer *peer;
+	const u8 *peer_addr;
+
+	WARN_ON(!rcu_read_lock_held());
+	if (!sta_addr)
+		return NULL;
+
+	if (list_empty(&ath12k_primary_peers))
+		return NULL;
+
+	list_for_each_entry_rcu(peer, &ath12k_primary_peers, primary_list) {
+		if (!peer)
+			continue;
+
+		peer_addr = peer->mlo ? peer->ml_addr : peer->addr;
+		if (!peer_addr)
+			continue;
+
+		if (ether_addr_equal(peer_addr, sta_addr)) {
+			found_peer = peer;
+			ath12k_dbg(NULL, ATH12K_DBG_PEER,
+				   "found primary peer : %pM\n", peer->addr);
+			break;
+		}
+	}
+
+	return found_peer;
+}
+
+void ath12k_peer_free(struct ath12k_peer *peer)
+{
+	if (!peer)
+		return;
+
+	ath12k_peer_remove_primary(peer);
+
+	call_rcu(&peer->rcu, ath12k_peer_rcu_callback);
+}
+
 struct ath12k_ml_peer *ath12k_ml_peer_find_by_id(struct ath12k_hw *ah,
 						 u16 ml_peer_id)
 {
@@ -308,7 +441,7 @@ void ath12k_peer_unmap_event(struct ath1
 	if (peer->sawf_ctx_peer.telemetry_peer_ctx)
 		ath12k_telemetry_peer_ctx_free(peer->sawf_ctx_peer.telemetry_peer_ctx);
 #endif
-	kfree(peer);
+	ath12k_peer_free(peer);
 	wake_up(&ab->peer_mapping_wq);
 
 exit:
@@ -553,7 +686,7 @@ void ath12k_peer_cleanup(struct ath12k *
 		if (peer->sawf_ctx_peer.telemetry_peer_ctx)
 			ath12k_telemetry_peer_ctx_free(peer->sawf_ctx_peer.telemetry_peer_ctx);
 #endif
-		kfree(peer);
+		ath12k_peer_free(peer);
 		ar->num_peers--;
 	}
 
--- a/drivers/net/wireless/ath/ath12k/peer.h
+++ b/drivers/net/wireless/ath/ath12k/peer.h
@@ -54,6 +54,7 @@ struct ath12k_dp_mon_peer_stats {
 
 struct ath12k_peer {
 	struct list_head list;
+	struct list_head primary_list;
 	u8 stats_disable;
 	struct net_device *dev;
 	struct ieee80211_sta *sta;
@@ -187,6 +188,13 @@ int ath12k_peer_send_assoc_vendor_respon
 bool ath12k_get_link_down(struct ath12k_sta *ahsta, int *link_going_down);
 int ath12k_peer_rhash_cleanup(struct ath12k *ar, u32 vdev_id, const u8 *addr);
 
+void ath12k_peer_add_primary(struct ath12k_peer *peer);
+void ath12k_peer_remove_primary(struct ath12k_peer *peer);
+struct ath12k_peer* ath12k_peer_find_primary_by_sta_addr(const u8 *sta_addr);
+void ath12k_peer_purge_primary_list(void);
+void ath12k_peer_purge_primary_list_per_ab(struct ath12k_base *ab);
+void ath12k_peer_free(struct ath12k_peer *peer);
+
 static inline
 enum nl80211_iftype ath12k_peer_get_peer_type(struct ath12k_peer *peer)
 {
